srml/offences/src/lib.rs

```diff
@@ -0,0 +1,147 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Offences Module
+//!
+//! Tracks reported offences
+
+// Ensure we're `no_std` when compiling for Wasm.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+mod mock;
+mod tests;
+
+use rstd::vec::Vec;
+use support::{
+	StorageDoubleMap, decl_module, decl_event, decl_storage, Parameter,
+};
+use sr_primitives::Perbill;
+use sr_staking_primitives::{
+	SessionIndex,
+	offence::{Offence, ReportOffence, TimeSlot, Kind, OnOffenceHandler, OffenceDetails},
+};
+
+/// Offences trait
+pub trait Trait: system::Trait {
+	/// The overarching event type.
+	type Event: From<Event> + Into<<Self as system::Trait>::Event>;
+	/// Full identification of the validator.
+	type IdentificationTuple: Parameter;
+	/// A handler called for every offence report.
+	type OnOffenceHandler: OnOffenceHandler<Self::AccountId, Self::IdentificationTuple>;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Offences {
+		/// A mapping between unique `TimeSlots` within a particular session and the offence `Kind` into
+		/// a vector of offending authorities.
+		///
+		/// It's important that we store all authorities reported for an offence for any kind and
+		/// timeslot since the slashing will increase based on the length of this vec.
```

Comment:
It's unlikely that we will have more than one offence at any given time-slot.
I understand that this PR isn't meant to have any rolling-window, but would you simply upgrade later by instead looking backwards + forwards (you probably need a notion of "now") from the time-slot of the actual offence?

Reply:
> It's unlikely that we will have more than one offence at any given time-slot.

A timeslot is a flexible notion that should inherently quantise to naturally group together typical cases of co-offending.

e.g. for Grandpa, the timeslot would be concerning one specific vote. If there were multiple offenders for that same vote then we slash them each more. If the offences were spread between different votes, then the danger is far lessened and they'd each get away with a lighter punishment. This is the case even if the voting still within a limited window of time.

For Babe and ImOnline, then timeslot could perhaps be the session/epoch quantisation.

For Parachains, then it would naturally be parachain block number.